src/Migrations/RefToInoutMigration.hack

```diff
@@ -129,11 +133,54 @@ final class RefToInoutMigration extends BaseMigration {
    *                           -1, inout $__unused_inout)
    */
   private static function optionalToRequired(
-    this::TNodes $nodes,
-    int $_arg_idx,
-    vec<string> $_previous_defaults = vec[],
+    this::TNodes $n,
+    int $inout_arg_idx,
+    vec<string> $previous_defaults = vec[],
   ): Script {
-    return $nodes['root'];  // TODO
+    $last_old_arg_idx = C\count($n['args']) - 1;
+    if ($last_old_arg_idx >= $inout_arg_idx) {
+      return self::refToInout($n, $inout_arg_idx);
+    }
+
+    // Figure out how many arguments with the provided default values we need
+    // to insert before the inout argument.
+    $default_cnt = $inout_arg_idx - $last_old_arg_idx - 1;
+    invariant(
+      $default_cnt <= C\count($previous_defaults),
+      'Need to add %d extra arguments but only %d values are available.',
+      $default_cnt,
+      C\count($previous_defaults),
+    );
+
+    $args_to_add = Vec\slice($previous_defaults, -$default_cnt)
+      |> Vec\map($$, $code ==> self::expressionFromCode($code));
+
+    // Add the new inout argument at the end.
+    $args_to_add[] = new DecoratedExpression(
+      new InoutToken(null, null),
+      new VariableToken(
+        NodeList::createMaybeEmptyList(vec[new WhiteSpace(' ')]),
+        null,
+        '$'.self::DUMMY_VARIABLE_NAME,
+      )
+    );
+
+    $new_call = add_arguments($n['root'], $n['call'], $args_to_add);
+
+    return $n['root']->replace($n['call'], $new_call)
+      |> prepend_statement($$, self::getDummyAssignment(), $new_call);
+  }
+
+  <<__Memoize>>
+  private static function getDummyAssignment(): IStatement {
+    return \HH\Asio\join(
```

Comment:
![image](https://user-images.githubusercontent.com/360927/64888422-d527db00-d61f-11e9-97a1-440ae55a84e1.png)

Reply:
It's memoized so it happens at most once.

It wouldn't be too hard to convert all the functions to async, but honestly I'm scared of what edge cases could surface if the order of execution of everything is no longer deterministic :/